src/renderer/atlas/AtlasEngine.api.cpp

```diff
@@ -0,0 +1,599 @@
+                default:
+                    fontFeatures.emplace_back(DWRITE_FONT_FEATURE{ tag, p.second });
+                    break;
+                }
+            }
+        }
+    }
+
+    std::vector<DWRITE_FONT_AXIS_VALUE> fontAxisValues;
+    if (!axes.empty())
+    {
+        fontAxisValues.reserve(axes.size() + 3);
+
+        // AtlasEngine::_recreateFontDependentResources() relies on these fields to
+        // exist in this particular order in order to create appropriate default axes.
+        fontAxisValues.emplace_back(DWRITE_FONT_AXIS_VALUE{ DWRITE_FONT_AXIS_TAG_WEIGHT, -1.0f });
+        fontAxisValues.emplace_back(DWRITE_FONT_AXIS_VALUE{ DWRITE_FONT_AXIS_TAG_ITALIC, -1.0f });
+        fontAxisValues.emplace_back(DWRITE_FONT_AXIS_VALUE{ DWRITE_FONT_AXIS_TAG_SLANT, -1.0f });
+
+        for (const auto& p : axes)
+        {
+            if (p.first.size() == 4)
+            {
+                const auto s = p.first.data();
+                switch (const auto tag = DWRITE_MAKE_FONT_AXIS_TAG(s[0], s[1], s[2], s[3]))
+                {
+                case DWRITE_FONT_AXIS_TAG_WEIGHT:
+                    fontAxisValues[0].value = p.second;
+                    break;
+                case DWRITE_FONT_AXIS_TAG_ITALIC:
+                    fontAxisValues[1].value = p.second;
+                    break;
+                case DWRITE_FONT_AXIS_TAG_SLANT:
+                    fontAxisValues[2].value = p.second;
+                    break;
+                default:
+                    fontAxisValues.emplace_back(DWRITE_FONT_AXIS_VALUE{ tag, p.second });
+                    break;
+                }
+            }
+        }
+    }
+
+    const auto previousCellSize = _api.fontMetrics.cellSize;
+    _resolveFontMetrics(fontInfoDesired, fontInfo, &_api.fontMetrics);
+    _api.fontFeatures = std::move(fontFeatures);
+    _api.fontAxisValues = std::move(fontAxisValues);
+
+    WI_SetFlag(_api.invalidations, ApiInvalidations::Font);
+
+    if (previousCellSize != _api.fontMetrics.cellSize)
+    {
+        _api.cellCount = _api.sizeInPixel / _api.fontMetrics.cellSize;
+        WI_SetFlag(_api.invalidations, ApiInvalidations::Size);
+    }
+
+    return S_OK;
+}
+CATCH_RETURN()
+
+void AtlasEngine::UpdateHyperlinkHoveredId(const uint16_t hoveredId) noexcept
+{
+}
+
+#pragma endregion
+
+void AtlasEngine::_resolveFontMetrics(const FontInfoDesired& fontInfoDesired, FontInfo& fontInfo, FontMetrics* fontMetrics) const
+{
+    auto requestedFaceName = fontInfoDesired.GetFaceName().c_str();
+    const auto requestedFamily = fontInfoDesired.GetFamily();
+    auto requestedWeight = fontInfoDesired.GetWeight();
+    auto requestedSize = fontInfoDesired.GetEngineSize();
+
+    if (!requestedFaceName)
+    {
+        requestedFaceName = L"Consolas";
+    }
+    if (!requestedSize.Y)
+    {
+        requestedSize = { 0, 12 };
+    }
+    if (!requestedWeight)
+    {
+        requestedWeight = DWRITE_FONT_WEIGHT_NORMAL;
+    }

```

Comment:
should we static/define these at the top?

Reply:
IMO This would hinder readability, because then you'd first have to check what that variable is defined as.

Reply:
Ah fine OK.